![Workshop-logo](./IM/LOGO_new.png)
# Applications of Genomics in Wildlife Conservation
Generating and analysing genome wide information from endangered populations can greatly assist population management. 
Manipulating genomic data, obtained from next-generation sequencing platforms, however, requires a set of skills in bioinformatics that often prove to be a major barrier. In this workshop you will learn the basics of next-generation sequencing data analysis for conservation

## Day 1 - Overview
Basic concepts of command line programing.
#### Session 1
- Accessing cloud server (ssh)
- Command Line Interface
- File System Structure
- Manipulating text files

#### Session 2
- Variables and arguments
- Shell Scripting
- Transferring files (sftp)
- Introduction to conda


## Day 2 - Overview
NGS data preparation for genomic analyses.

#### Session 1
- Introduction to commonly used bioinformatic file formats
- Assessing the quality of sequencing results
- Aligning reads to reference genomes

#### Session 2
- Manipulating alignment files (bam)
- Estimating coverage, endogenous content, and percentage of duplicates

#### Session 3
- Parallelization strategies
- Variant calling

## Day 3 - Overview
Analysing genetic variation.

#### Session 1
- Understanding the .vcf file
- Calculating segregating sites, average number of pairwise differences, heterozygosity
- Hypothesis for the observed genetic diversity
- Discuss how to test these hypothesis

#### Session 2
- Generating phylogenetic trees
- Population structure with PCA
- Investigating population admixture

#### Session 3
- Visualising trees with iTOL
- Using R to visual PCA and ADMIXTURE results
- Discussion of Babirusa population structure

## Day 4 - Overview
Demographic analysis, inbreeding & load.

#### Session 1
- Prepping input files for PSMC
- Running and discussing PSMC

#### Session 2
- Running observational ROH detection method using PLINK
- Using R to visualise ROH observation-based and model-based results
- Discussion of different areas inbreeding extent

#### Session 3
- Running custom pipeline to calculate mutational load per individual
- Using R to plot mutational load per area
- Discussion of burden of mutation in different areas

## Day 5 - Overview
Advanced population genetics & Metabarcoding for biodiversity.

#### Session 1
- Introduction to f-statistics
- Detecting admixture
- Quatifying admixture proportions.

#### Session 2
- Overview of analysis of biodiversity with metabarcoding data
- Run through of DADA2 pipeline in R

#### Session 3
- Calculating biodiversity metrics from OTU/ASVs in R






